src/hl/dataset.rs

```diff
@@ -140,728 +168,1112 @@ impl Dataset {
+        T: H5Type,
+        D: ndarray::Dimension,
+    {
+        DatasetBuilderData {
+            builder: self.builder,
+            data: data.into(),
+            type_desc: type_desc.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+/// A dataset builder with the type known
+pub struct DatasetBuilderEmpty {
+    builder: DatasetBuilderInner,
+    type_desc: TypeDescriptor,
+}
+
+impl DatasetBuilderEmpty {
+    pub fn shape<S: Into<Extents>>(self, extents: S) -> DatasetBuilderEmptyShape {
+        DatasetBuilderEmptyShape {
+            builder: self.builder,
+            type_desc: self.type_desc,
+            extents: extents.into(),
+        }
+    }
+    pub fn create<'n, T: Into<Maybe<&'n str>>>(self, name: T) -> Result<Dataset> {
+        self.shape(()).create(name)
+    }
+}
+
+#[derive(Clone)]
+/// A dataset builder with type and shape known
+pub struct DatasetBuilderEmptyShape {
+    builder: DatasetBuilderInner,
+    type_desc: TypeDescriptor,
+    extents: Extents,
+}
+
+impl DatasetBuilderEmptyShape {
+    pub fn create<'n, T: Into<Maybe<&'n str>>>(&self, name: T) -> Result<Dataset> {
+        h5lock!(self.builder.create(&self.type_desc, name.into().into(), &self.extents))
+    }
+}
+
+#[derive(Clone)]
+/// A dataset builder with type, shape, and data known
+pub struct DatasetBuilderData<'d, T, D> {
+    builder: DatasetBuilderInner,
+    data: ArrayView<'d, T, D>,
+    type_desc: TypeDescriptor,
+}
+
+impl<'d, T, D> DatasetBuilderData<'d, T, D>
+where
+    T: H5Type,
+    D: ndarray::Dimension,
+{
+    pub fn create<'n, N: Into<Maybe<&'n str>>>(&self, name: N) -> Result<Dataset> {
+        ensure!(
+            self.data.is_standard_layout(),
+            "input array is not in standard layout or is not contiguous"
+        ); // TODO: relax this when it's supported in the writer
+        let extents = Extents::from(self.data.shape());
+        let name = name.into().into();
         h5lock!({
-            let defined: *mut H5D_fill_value_t = &mut H5D_fill_value_t::H5D_FILL_VALUE_UNDEFINED;
-            let dcpl_id = self.dcpl_id()?;
-            h5try!(H5Pfill_value_defined(dcpl_id, defined));
-            match *defined {
-                H5D_fill_value_t::H5D_FILL_VALUE_ERROR => fail!("Invalid fill value"),
-                H5D_fill_value_t::H5D_FILL_VALUE_UNDEFINED => Ok(None),
-                _ => {
-                    let datatype = Datatype::from_type::<T>()?;
-                    let mut value = mem::MaybeUninit::<T>::uninit();
-                    h5try!(
-                        H5Pget_fill_value(dcpl_id, datatype.id(), value.as_mut_ptr() as *mut _,)
-                    );
-                    Ok(Some(value.assume_init()))
-                }
+            let dtype_src = Datatype::from_type::<T>()?;
+            let dtype_dst = Datatype::from_descriptor(&self.type_desc)?;
+            // TODO: soft conversion? hard? user-specifiable?
+            dtype_src.ensure_convertible(&dtype_dst, Conversion::Soft)?;
```

Comment:
Conversion: this would be quite easy to do, no? By just adding the same API to the builder as in `Container`:

```rust
    /// Set maximum allowed conversion level.
    pub fn conversion(mut self, conv: Conversion) -> Self {
        self.conv = conv;
        self
    }

    /// Disallow all conversions.
    pub fn no_convert(mut self) -> Self {
        self.conv = Conversion::NoOp;
        self
    }
```

Reply:
Done